PrebidMobileTests/RenderingTests/Tests/ParameterBuilderTests/PrebidParameterBuilderTest.swift

```diff
@@ -784,6 +784,24 @@ class PrebidParameterBuilderTest: XCTestCase {
             XCTAssertEqual(imp.extGPID, gpid)
         }
     }
+    
+    func testArbitraryORTBParams() {
+        let gpid = "/12345/home_screen#identifier"
+        let globalOrtb = ["arbitraryparam1": "param1", "arbitraryparam2": 2, "arbitraryparam3": false] as [String : Any]
+        let impOrtb = ["arbimpparam1": "impparam1", "arbitraryparam2": 2, "arbitraryparam3": false] as [String : Any]
+        let adUnit = AdUnit(configId: "test", size: CGSize.zero, adFormats: [.banner])
+        adUnit.setGPID(gpid)
+        adUnit.setGlobalOrtbObject(globalOrtb)
+        adUnit.setImpOrtbObject(impOrtb)
```

Comment:
Hi @jsligh, @bretg ! 

The current implementation looks good, and it can address the needs described in the ticket. But also, it seems too fragile and publisher-dependent in the long-term perspective. 

For example, by providing an API to set the Global object and Imp object independently sooner or later, we will get a situation where the publisher mixes these approaches, and the final request will have the wrong structure. 

Also, correct me if I am wrong, but from what I see in the `PBMORTBAbstractTest.swift`, if a publisher provides, for example, the same `ext` object via the `ortbObject` and via `AdUnit` API, the filed will be duplicated in the final bid request. Instead, the data provided in ortbObject should override any other data. At the same time, we can't just ignore data received from API fields because they can be set by 3p services, like identity providers, so the SDK update may lead to losing data in the bid request silently.

Also, there are no changes for Rendering and Mediation API in this PR. 

As I told you from the very beginning, it is an extremely complex task that assumes a lot of design work first. So, instead of searching for implementation gaps, let's design the implementation spec first, and if it covers all the objectives, is robust for integration, and is unambiguous for utilization, we'll update the PR accordingly. 

I can provide you with a draft version early next week so we can discuss it. 

cc: @alexsavelyev , @mmullin

Reply:
Thanks for the thoughts Yuriy. My take is based on the experience we've had with Prebid.js:

- ORTB is a standard. Developers have to understand it anyhow in order to debug.
- Things in our industry change too quickly for an API to keep up
- Therefore, instead of trying to design a new API function everytime there's a new feature, letting developers have direct access to ORTB is better in every way: smaller API footprint, more robust to continuing evolution, developers have less to learn.

I agree that the "rendering API" also needs this functionality, though as we discussed earlier today, I frankly have issues with a core assumption of that API -- that Prebid can always choose the winner. I'm going to open a separate issue to discuss that.

Reply:
I agree with all points and statements except: 

> Developers have to understand it anyhow in order to debug.

With full respect, but from all our experience in mobile development on both sides - publishers and the ad tech industry, I should say that developers of mobile apps don't have a knowledge of ORTB. 

Open RTB, as usual, is used in s2s cases. In the mobile world, the ad request (SDK --> SSP) has a custom format depending on each particular implementation. App developers, in their own turn, work only with the API of each particular SDK that provides a certain set of parameters to tune. This API and integration documentation is the final frontier of their knowledge of ad tech and what is happening behind the ad requests to the SSP (or other platforms).

Of course, with exceptions when publishers have an AdOps department that directly works on monetization. These guys have a great understanding of the ecosystem. 

So, in order to prevent logical errors, it makes sense to restrict the ability to shoot themselves in the foot as much as possible.

Reply:
@bretg @jsligh please review the proposal. For now, it is in the doc so you can easily edit it. Once everything is aligned we can copy the text here: 

https://docs.google.com/document/d/1cRWDNO6i7wLyQdQA9ZzmK0QumIr2vvC40r5GhHhOSz4/edit?usp=sharing

Sorry if the intro sections are too long. I tried to highlight the context and pitfalls that impact the proposed design. 

cc: @alexsavelyev @mmullin